[specific ci=1-39-Docker-Stats] Add network and disk stats to docker stats

cgtexmex opened the pull request:
Docker stats now reports network and disk stats. Additionally,
stats collection refactored to only run when there are active
subscriptions.

Fixes #4422, #3962, #5746

cgtexmex commented:
robot tests are in progress, but the code is complete

dougm commented on lib/portlayer/metrics/container.go:
```diff
@@ -66,113 +62,103 @@ type MemoryMetrics struct {
+	// loop on the channel
+	go func() {
+		for {
+			select {
```
Stop() doesn't close the ticker channel, don't you need another case in here to break the loop / return from the go func?

cgtexmex replied:
Yes -- I thought that since Stop() "released the ticker resources" that it would close...but that's not the case.  I've added logic (locally) to break out of the loop -- I'll push up for review in a few.

cgtexmex commented:
@dougm now correctly exiting from the polling/collection loop

mhagen-vmware commented on lib/apiservers/engine/backends/convert/stats.go:
```diff
@@ -316,3 +419,40 @@ func (cs *ContainerStats) currentCPU() {
+// if 1 is provided 'a' will be returned, 2 == 'b', 3 == 'c', etc..
+func toChar(i int) string {
+	return abc[i-1 : i]
+}
```
I can't see where this is used in this PR? Simple grep at least doesn't show anything..

cgtexmex replied:
Yeah..it's not used...I thought originally I would need to use it...I'll remove.

mhagen-vmware commented on lib/apiservers/engine/backends/convert/stats.go:
```diff
@@ -316,3 +419,40 @@ func (cs *ContainerStats) currentCPU() {
+	return uint64(minor)
+}
+
+func createBlkioStatsEntry(minor uint64, op string, value uint64) types.BlkioStatEntry {
```
should this have some error handling?  minor and string are generally covered by your own code - i.e. minor is either going to be valid or 0 and string is hard coded to read/write/total in all places, but value does not appear to be guaranteed to be a uint64 necessarily in the case of an error?

cgtexmex replied:
the value will always be uint64...IMO no error handling needed

mhagen-vmware commented on lib/portlayer/metrics/container.go:
```diff
@@ -224,117 +215,200 @@ func (cc *collectorVM) collect(currentSubs map[types.ManagedObjectReference]*sub
+	if err != nil {
+		// I don't expect this to ever happen, but if it does log and don't publish
+		log.Errorf("metrics failed to convert the container(%s) device id to an int - value(%#v): %s", containerID, instance, err)
+		return converted, err
```
I am not familiar enough with golang Atoi, but since you move this here and out of the switch statement above - does Atoi at least populate converted with 0/-1 or something sensible so that we are guaranteed we are not returning some totally invalid value?

cgtexmex replied:
If it fails the int will be zero -- but if it fails we'll return the err and we'll skip that device, so we won't report.

I moved it out of the switch as I thought I'd be using it in other areas (disk, net & not just CPU), but didn't need it for those...

sgairo commented on lib/apiservers/engine/backends/convert/stats.go:
```diff
@@ -42,6 +44,11 @@ type ContainerStats struct {
 	currentMetric *metrics.VMMetrics
 
+	// disk & net stats are accumulated during the life of the
+	// subscription. These maps will assist in that accululation
```
"accumulation" :)

cgtexmex replied:
I like my spelling better... ;-)

dougm commented on lib/portlayer/metrics/container.go:
```diff
@@ -66,113 +62,117 @@ type MemoryMetrics struct {
-
-	return ch, nil
+	cc.timer.Stop()
+	cc.stopper <- true
```
You can just close the channel here, won't be used after this anyhow.  The golang nerds would also tell you to use `chan struct{}` instead of `chan bool`, since it uses less memory.

cgtexmex replied:
You can't close the channel as it's receive only.  (https://play.golang.org/p/NACq0RYGGd)

I originally had as `chan struct{}` for that reason, but then changed to bool...

dougm replied:
I meant close the cc.stopper channel, not the ticker.C

cgtexmex replied:
ah..yeah..I could...was just avoiding the `make` on each Start...regardless I'm switching to an empty struct and will be closing...so we can all be happy ;-)

caglar10ur commented on lib/apiservers/engine/backends/convert/stats.go:
```diff
@@ -42,6 +43,11 @@ type ContainerStats struct {
+	// disk & net stats are accumulated during the life of the
+	// subscription. These maps will assist in that accumulation.
+	diskStats map[string]metrics.VirtualDisk
+	netStats  map[string]metrics.Network
```
Do we drop the ContainerStats when the container gets removed? Also how about restarts, do we just keep adding or somehow start from zero values?

cgtexmex replied:
We drop the stats when the stats stream (subscription) ends.  That could be via the client or the API.  If the interaction is without streaming (`docker stats --no-stream`) then there's no accumulation and only a point in time look at disk and network.

hickeng commented on lib/portlayer/metrics/subscription.go:
```diff
@@ -0,0 +1,180 @@
+
+	// convert instance to key - we are expecting regular failures, so no logging
+	// or returning of error
+	key, err := strconv.Atoi(instance)
```
do you have an example of an `instance` that shouldn't be translated we can add to doc?
I'm not clear on what would be passed in that's not translated - I also cannot see where this method is called.

hickeng commented on lib/portlayer/metrics/subscription.go:
```diff
@@ -0,0 +1,180 @@
+// subscription is a 1:1 relationship to a container and a
+// 1:M relationship to subscribers
+type subscription struct {
+	c *exec.Container
```
Is it viable for this to use `vm.VirtualMachine` instead of exec.Container? I'm wondering if this needs to be in the portlayer or if it will happily sit in `pkg/vsphere`/potential inclusion into govmomi.

cgtexmex replied:
Yes we could use `vm.VirtualMachine` -- want me to do that now or later?

hickeng commented on lib/portlayer/metrics/subscription.go:
```diff
@@ -0,0 +1,180 @@
+}
+
+// newContainerSubscription is a helper func to convert the interface to a subscription
+func newContainerSubscription(contain interface{}) (*subscription, error) {
```
Also couldn't find a call - in this PR or searching current master.

cgtexmex commented on lib/apiservers/engine/backends/convert/stats.go:
```diff
@@ -235,6 +247,95 @@ func (cs *ContainerStats) ToContainerStats(current *metrics.VMMetrics) (*types.S
+// Supported stats are io_service_bytes_recursive and io_serviced_recursive, so bytes and iops
+// during the stats subscription
+//
+// TODO: Currently disk assumes a single scsi device.  Multiple scsi devices will need
```
Controllers, not devices...updated comment forthcoming...

cgtexmex commented on lib/apiservers/engine/backends/convert/stats.go:
```diff
@@ -205,10 +213,8 @@ func (cs *ContainerStats) Listen() *io.PipeWriter {
-			// we've already got this as current, so skip and wait for the
-			// next sample
+		// do we have the same metric or has the metric not been initialized?
+		if cs.currentMetric.SampleTime.Equal(current.SampleTime) || current.SampleTime.IsZero() {
```
yes - at the end of a stats call there was an intermittent issue where a final metric would be produced, but it wasn't completely initialized.   The easiest way to tell is the fact that the sampleTime wasn't set.

cgtexmex commented on lib/apiservers/engine/backends/convert/stats.go:
```diff
@@ -316,3 +417,31 @@ func (cs *ContainerStats) currentCPU() {
+	// retrieve the last char
+	last := string(name[len(name)-1])
+	// the int suffix should always be last
+	minor, err := strconv.Atoi(last)
```
when creating this I apparently switched controller maximum w/disk maximum...so this was coded to having fewer than 10 disks.  I'll need to parse the string differently.  FTR the names look like this: `scsi0:0`

cgtexmex commented on lib/apiservers/engine/backends/convert/stats.go:
```diff
@@ -75,6 +81,8 @@ func NewContainerStats(config *ContainerStatsConfig) *ContainerStats {
 		curDockerStat: &types.StatsJSON{},
 		totalVCHMhz:   uint64(config.VchMhz),
 		dblVCHMhz:     uint64(config.VchMhz * 2),
```
It is not related to vCPU count.  I have a comment as to why this is the case when the variable is used:  https://github.com/cgtexmex/vic/blob/stats-4422/lib/apiservers/engine/backends/convert/stats.go#L402

cgtexmex commented on lib/portlayer/exec/base.go:
```diff
@@ -97,6 +98,19 @@ func (c *containerBase) VMReference() types.ManagedObjectReference {
 }
 
+// Devices will return a VirtualDeviceList
+func (c *containerBase) Devices() (object.VirtualDeviceList, error) {
```
Yes...I'll get it added...

cgtexmex commented:
@hickeng I've moved the performance collector to `pkg/vsphere/performance` thus decoupling from exec and positioning for the gateway.  Additionally I've added tracing to the flow.  Let me know if you have other questions / concerns.

hickeng commented on lib/portlayer/metrics/metrics.go:
```diff
@@ -34,18 +38,19 @@ var (
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	return fmt.Sprintf("type %s is not supported by metrics", t.Name())
```
"%T" didn't provide sufficient detail?

cgtexmex replied:
I've changed it...to just the type...so `fmt.Sprintf("type %s is not supported by metrics", t)`

hickeng commented on lib/portlayer/metrics/metrics.go:
```diff
@@ -67,13 +72,36 @@ func Init(ctx context.Context, session *session.Session) error {
+func (s *super) Subscribe(op trace.Operation, subscriber interface{}) (chan interface{}, error) {
+	switch subscriber.(type) {
+	case *exec.Container:
+		container := subscriber.(*exec.Container)
```
type cast is unneeded in a type switch. It's automatic through Go magic.

cgtexmex replied:
Not sure that I'm following.  I'm going to call methods if it's a container, so I have to cast it.  What am I missing?

cgtexmex replied:
done `switch sub := subscriber.(type){...}`

hickeng commented on lib/portlayer/metrics/metrics.go:
```diff
@@ -67,13 +72,36 @@ func Init(ctx context.Context, session *session.Session) error {
+func (s *super) Unsubscribe(op trace.Operation, subscriber interface{}, ch chan interface{}) {
+	switch subscriber.(type) {
+	case *exec.Container:
+		s.vms.Unsubscribe(op, subscriber.(*exec.Container).VMReference(), ch)
```
follow the code structure/patter in `Subscribe`

hickeng commented on pkg/vsphere/performance/vm_collector.go:
```diff
@@ -0,0 +1,407 @@
+						switch v.Name {
+						case "net.bytesRx.average":
+							// perfManager returns kbps -- convert to sum of bytes
+							sum := summation(v.Value[s]) * 1024
```
`kbps * 1024` translates to `bps` - is `Rx.Bytes` actually `Rx.BytesPerSecond` or is there an unstated assumption that the interval is 1s?

cgtexmex replied:
This is total bytes for the interval.  The `summation` func will multiply the `kbps` by the `interval` and then we multiply that by `1024` to get total bytes.
